proto/identity/associations/association.proto

```diff
@@ -0,0 +1,78 @@
+syntax = "proto3";
+
+package xmtp.identity.associations;
+
+import "identity/associations/signature.proto";
+
+option go_package = "github.com/xmtp/proto/v3/go/identity/associations";
+option java_package = "org.xmtp.proto.identity.associations";
+
+// The first entry of any XID log, which is considered valid iff the XID can be
+// deterministically derived from the address and nonce.
+// The recovery address defaults to the initial_associated_address unless
+// there is a subsequent ChangeRecoveryAddress in the log.
+message CreateXid {
+  string xid = 1;
+  uint32 nonce = 2;
+  string initial_associated_address = 3;
+}
+
+message AddAssociation {
+  enum AssociatedRole {
+    ASSOCIATED_ROLE_UNSPECIFIED = 0;
+    ASSOCIATED_ROLE_INSTALLATION = 1;
+    ASSOCIATED_ROLE_ADDRESS = 2;
+    ASSOCIATED_ROLE_LEGACY_KEY = 3;
+  }
+
+  string xid = 1;
+  uint64 client_timestamp_ns = 2;
+  uint32 nonce = 3;
```

Comment:
If the nonce is meant to be globally unique, a string GUID might be better.

However, I think it's better if it is not required to be globally unique - for replay protection, we really just need it to be unique across the (existing member, new member, role) tuple. If that's the case, could we use the `client_timestamp_ns` as effectively the nonce? It has the benefit that unintentional nonce reuse is very hard to do, as well as the fact that it displays better in association text to the user.

And instead of storing the `revoked_nonce` below, I wonder if it makes sense to store a hash of the whole association instead. That way, it is clear that there is no assumption of global uniqueness for the nonce.

Reply:
The nonce only needs to be locally unique to the XID. Timestamp would work, but I'm imagining it's going to get generated somewhere inside our SDK so I'm not sure it matters.

> And instead of storing the revoked_nonce below, I wonder if it makes sense to store a hash of the whole association instead. That way, it is clear that there is no assumption of global uniqueness for the nonce.

Works for me

Reply:
Okay, so I removed the nonce. For replay protection, I'd suggest we detect replays using the hash of the whole association - it is up to the client to ensure that the client timestamp will be different between different signature attempts if they want their signature to be accepted.

Reply:
@neekolas I just realized that protobuf serialization is not guaranteed to be consistent, hence we can't use a hash of the protobuf bytes for replay protection. Maybe a hash of the signature bytes would work? If not, maybe we could manually construct a tuple with our own serialization method

Reply:
In my implementation I've just baked a function `hash()` into the trait for each log entry, so each type can define an appropriate hash function.

I haven't worked out the signature text for all the log entries, but I _think_ that's the only input needed to get replay prevention. Anything important should go into the signature text.

If that's the case I'll remove the `hash()` function and just use the text.